treeherder/model/management/commands/cycle_data.py

```diff
@@ -442,6 +443,89 @@ def _find_ideal_chunk_size(self) -> int:
         return len(older_perf_data_rows) or self._chunk_size
 
 
+class StalledDataRemoval(RemovalStrategy):
+    """
+    Removes `performance_signature` rows that haven't
+    been updated in the last 4 months.
+
+    Also removes all their associated `performance_datum`
+    & `performance_alert` rows.
+    """
+
+    def __init__(self, chunk_size: int):
+        self._cycle_interval = timedelta(days=120)
+        self._chunk_size = chunk_size
+        self._max_timestamp = datetime.now() - self._cycle_interval
+        self._target_signature = None
+        self._removable_signatures = None
+
+    @property
+    def target_signature(self) -> PerformanceSignature:
+        try:
+            if self._target_signature is None:
+                self._target_signature = self.removable_signatures.pop()
```

Comment:
Potential duplicated code in `__lookup_new_signature` method: Perhaps adding a setter for target_signature and, in method `remove` after line ` self.__attempt_remove(using)` you could do `self.target_signature = None`.

Reply:
There is some kind of duplication, because of the way I pop() from `self.removable_signatures` in 2 code sections.
I preferred it this way so that there's a distinction between having no removable signatures from the start **vs** depleting them in the end (which is the general expected behavior). It allows more obvious logging & easier troubleshooting.

Setting `self.target_signature = None` isn't a good option, as it will re target the next signature, without removing all the data points from the previous one.